File: engineering/LeetCode/2138.divide-a-string-into-groups-of-size-k.ts
/*
 * @lc app=leetcode id=2138 lang=typescript
 *
 * [2138] Divide a String Into Groups of Size k
 */

// @lc code=start
/**
 * Time complexity: O(s / k)
 * Space complexity: O(s + k - s % k)
 */
function divideString(s: string, k: number, fill: string): string[] {
    if (s.length % k !== 0) { s += fill.repeat(k - s.length % k); }
    const res: string[] = [];
    let i = 0;
    while (i <= s.length - k) {
        res.push(s.substring(i, i + k));
        i += k;
    }
    return res;
};
// @lc code=end

File: engineering/LeetCode/2138.Divide-a-String-Into-Groups-of-Size-k.py
"""
https://leetcode.com/problems/divide-a-string-into-groups-of-size-k/description/

Time complexity: O(n), n = len(s)
Space complexity: O(n)
"""
class Solution:
    def divideString(self, s: str, k: int, fill: str) -> List[str]:
        res, cur = [], ''

        for c in s:
            cur += c
            if len(cur) == k:
                res.append(cur)
                cur = ''
        
        if len(cur) != 0:
            cur += fill * (k - len(cur))
            res.append(cur)
        return res
    